lib/OpenQA/Worker/Cache.pm

```diff
@@ -372,144 +285,114 @@ sub _update_asset_last_use {
 
 sub update_asset {
     my ($self, $asset, $etag, $size) = @_;
-
-    my $dbh = $self->dbh;
-
-    my $sql = "UPDATE assets set filename =?, etag =? , size = ?, last_use = strftime('%s','now') where filename = ?;";
-    my $res;
-
     eval {
-        $dbh->do('BEGIN EXCLUSIVE');
-        my $sth = $dbh->prepare($sql);
-        $sth->bind_param(1, $asset);
-        $sth->bind_param(2, $etag);
-        $sth->bind_param(3, $size);
-        $sth->bind_param(4, $asset);
-
-        $res = $sth->execute;
-        $dbh->commit;
+        my $tx  = $self->dbh->db->begin('exclusive');
+        my $sql = q(UPDATE assets set etag =? , size = ?, last_use = strftime('%s','now') where filename = ?;);
+        $self->dbh->db->query($sql, $etag, $size, $asset);
+        $tx->commit;
     };
-
-    $self->cache_real_size($self->cache_real_size + $size);
-
     if ($@) {
         log_error "Update asset $asset failed. Rolling back $@";
-        eval { $dbh->rollback };
-        log_error "Rolling back failed: $@";
     }
     else {
         log_info "CACHE: updating the $asset with $etag and $size";
     }
 
-    return !!0 if !defined $res || $res == 0 || $@;
-    return !!1 if defined $res && $res == 1;
+    my $result = $self->dbh->db->select('assets', [qw(etag size last_use filename)], {filename => $asset})->arrays;
+
+    return !!0 if $result->size == 0 || $@;
+    $self->increase($size) and return !!1 if $result->size == 1;
 }
 
 sub purge_asset {
     my ($self, $asset) = @_;
-    my $sql = "DELETE FROM assets WHERE filename = ?";
-    my $dbh = $self->dbh;
-
     eval {
-        $dbh->do("BEGIN EXCLUSIVE");
-        $dbh->prepare($sql)->execute($asset) or die $dbh->errstr;
-        $dbh->commit or die $dbh->errstr;
-        unlink($asset) or eval { log_error "CACHE: Could not remove $asset" if -e $asset };
-        log_debug "CACHE: removed $asset";
+        my $tx = $self->dbh->db->begin();
+        $self->dbh->db->delete('assets', {filename => $asset});
+        $tx->commit;
```

Comment:
Another case where you need to hold on to the db handle instead of calling `$self->dbh->db` repeatedly.